### ISSUE TITLE
Does NATS server have plan to support http3(quic protocol)?

### ISSUE THREAD

**Issue by xieyuschen** [nats-io/nats-server#3140]

**Comment by Jarema** [comment]
Hey!
Quic protocol support is on the roadmap for the second half of the year.

**Comment by xieyuschen** [comment]
> Hey!
> Quic protocol support is on the roadmap for the second half of the year.

wow so cool, thanks. Are there any opportunities for me to get involved?

**Comment by derekcollison** [comment]
Of course. Our tentative plan is to support it for client connections and leafnodes, similar to how we did websockets.

**Comment by xieyuschen** [comment]
> Of course. Our tentative plan is to support it for client connections and leafnodes, similar to how we did websockets.

thanks:)

**Comment by xieyuschen** [comment]
> Of course. Our tentative plan is to support it for client connections and leafnodes, similar to how we did websockets.

Is it ok to make a pr before July or you just consider to review this feature after June?

**Comment by derekcollison** [comment]
Happy to review a PR!

**Comment by gedw99** [comment]
Updates: the golang core team are looking at integrating quic into the core code base of golang.

it’s a very long road because the testing and net api will need to be adapted to the go teams conventions. But it’s nice to see this getting started.

**Comment by derekcollison** [comment]
That would make it a no brainer for us.

**Comment by jdrews** [comment]
Here's the proposal for adding quic to core go: https://github.com/golang/go/issues/58547

And the draft API: https://go-review.googlesource.com/c/net/+/468575

**Comment by jdrews** [comment]
Lots of great progress on adding quic to core go. More to do, but it's shaping up in go/net master branch already!
https://go.googlesource.com/net/+/refs/heads/master/internal/quic/

https://go-review.googlesource.com/c/net/+/468402/4

**Comment by wbaileyisla** [comment]
Any updates? It seems like everything was merged?

**Comment by derekcollison** [comment]
We are considering for 2.11.

**Comment by neilalexander** [comment]
We are still thinking and talking about QUIC support, but it won't be in 2.11 I'm afraid.

**Comment by wbaileyisla** [comment]
I will literally send you a new puppy or kitten if this is moved up

**Comment by ripienaar** [comment]
@wbaileyisla what are the benefits you aim to achieve by using QUIC and for what kind of connection exactly?

**Comment by wbaileyisla** [comment]
Wrpc is getting a lot of play lately, and we are using nats with wasmcloud and wasi/wrpc. If you added quic support we could continue to do amazing things with wasmcloud etc (as well as anyone who uses nats). Auth all the way to connection id would be fantastic. There aren't enough amazing things I can say about nats.. but quic support would really take off the restrictor plate and show 'em what the ole red dragon can do.

**Comment by ktalg** [comment]
@neilalexander Hi!
We’re currently working on adding QUIC support to nats-server to improve gateway communication efficiency in weak network environments.

Right now, we’re reviewing the gateway logic and planning some changes to the basic communication model with a few tweaks. The idea is to use multiple unidirectional streams instead of a single TCP connection:

1. For non-RMSG messages (like info, rs+, rs-, etc.), we’d use numbering to ensure message order.
2. For RMSG messages, we’re thinking of hashing the subject and sending those through separate streams.

This way, messages for the same subject stay ordered, but different messages won’t block or interfere with each other. Internal messages (non-RMSG) will also stay ordered.

We’re trying to keep these changes as minimal as possible and only apply them to gateway connections. Does this approach sound workable to you?

**Comment by derekcollison** [comment]
Looping in @kozlovic who understands the gateway code quite well.

**Comment by kozlovic** [comment]
@ktalg We have been plagued with propagation interest races since the start of the gateway project, due to the fact that the subscription (no)interest is sent to  inbound gateway connections while data is sent to the outbound connections. We have code in place to try mitigate that - and for the most cases it works well, but I have started looking (* note 1) at changing that so that the subscription (no)interest would also go through the outbound connection (there is still a possible race within the origin cluster if a message comes from the gateway and "lands" on a server that did not receive yet the subscription interest from the cluster where it originated), and your approach would then re-introduce this kind of subscription propagation race.

What you say is that the non-RMSG would be ordered, but that does not guarantee that an RMSG would not make it before a RS+ for instance, which could cause the server across the gateway - in the case of a request/reply situation, or simply a case where the receiving app generates another message on a different subject in response to it - to drop the "response" message because the RS+ - although it has been sent first - could be received after, because it used a different stream.

At the very least, I would therefore recommend to "pin" an account to a stream. But if you are using a single account and already experience a traffic bottleneck, then that would not help.

* Note 1: I started looking at changing the way the subscription interest is stored locally (currently the gateway has its own sublist) and how it is propagated, and was planning to also add multiple connections (possibly with compression support), very similar to what I did for the routes, but again, that means binding an account to a specific TCP connection. And I did only exploratory work at this point, nothing concrete at all, so no commitment on my part that this will ever see light ;-), and not related at all to QUIC.

**Comment by ZeroHans4** [comment]
Is this still on the horizon?
